File: methods/Recovery/Module6_Fragment_Recovery_Production_Baseline_v0.1.0/module6_fragment_recovery/src/signatures.cpp
#include "fragment/signatures.hpp"
#include <array>
#include <cstring>
namespace frag{
FileType parse_type(const std::string&s){if(s=="pdf")return FileType::PDF;if(s=="jpeg"||s=="jpg")return FileType::JPEG;if(s=="png")return FileType::PNG;if(s=="zip")return FileType::ZIP;return FileType::Unknown;}
static bool match(FileType t,const std::byte*p,size_t n){if(t==FileType::PDF)return n>=5&&std::memcmp(p,"%PDF-",5)==0;if(t==FileType::JPEG)return n>=3&&std::to_integer<unsigned char>(p[0])==0xFF&&std::to_integer<unsigned char>(p[1])==0xD8&&std::to_integer<unsigned char>(p[2])==0xFF;if(t==FileType::PNG)return n>=8&&std::memcmp(p,"\x89PNG\r\n\x1a\n",8)==0;if(t==FileType::ZIP)return n>=4&&std::memcmp(p,"PK\x03\x04",4)==0;return false;}
std::vector<Fragment> find_anchors(ISource&s,FileType t,std::uint64_t block,std::uint64_t max_scan){std::vector<Fragment>o;std::vector<std::byte>b(4*1024*1024);std::uint64_t end=std::min(max_scan,s.size());for(std::uint64_t p=0;p<end;){auto n=std::min<std::uint64_t>(b.size(),end-p);if(!s.read_at(p,std::span<std::byte>(b.data(),(size_t)n)))break;for(std::uint64_t i=0;i<n;i++)if(match(t,b.data()+i,(size_t)(n-i))){Fragment f{};f.id=o.size()+1;f.physical=p+i;f.length=block;f.file_offset=0;f.type=t;f.anchor_score=65;f.evidence.push_back("format_start_signature");o.push_back(f);}p+=n;}return o;}
}

File: repo/DriveWipe-main/DriveWipe-main/crates/drivewipe-core/src/forensic/signatures.rs
use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::io::RawDeviceIo;

/// A detected file signature at a specific offset.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileSignatureHit {
    /// Offset in bytes where the signature was found.
    pub offset: u64,
    /// Type of file detected.
    pub file_type: String,
    /// The magic bytes that were matched.
    pub magic_hex: String,
    /// Confidence level (high, medium, low).
    pub confidence: String,
}

/// Known file signatures (magic bytes).
struct FileSignature {
    name: &'static str,
    magic: &'static [u8],
    offset: usize,
}

const SIGNATURES: &[FileSignature] = &[
    FileSignature {
        name: "JPEG",
        magic: b"\xFF\xD8\xFF",
        offset: 0,
    },
    FileSignature {
        name: "PNG",
        magic: b"\x89PNG\r\n\x1A\n",
        offset: 0,
    },
    FileSignature {
        name: "PDF",
        magic: b"%PDF",
        offset: 0,
    },
    FileSignature {
        name: "ZIP/DOCX/XLSX",
        magic: b"PK\x03\x04",
        offset: 0,
    },
    FileSignature {
        name: "GIF87a",
        magic: b"GIF87a",
        offset: 0,
    },
    FileSignature {
        name: "GIF89a",
        magic: b"GIF89a",
        offset: 0,
    },
    FileSignature {
        name: "EXE/DLL",
        magic: b"MZ",
        offset: 0,
    },
    FileSignature {
        name: "ELF",
        magic: b"\x7FELF",
        offset: 0,
    },
    FileSignature {
        name: "RAR",
        magic: b"Rar!",
        offset: 0,
    },
    FileSignature {
        name: "7z",
        magic: b"7z\xBC\xAF\x27\x1C",
        offset: 0,
    },
    FileSignature {
        name: "BMP",
        magic: b"BM",
        offset: 0,
    },
    FileSignature {
        name: "GZIP",
        magic: b"\x1F\x8B",
        offset: 0,
    },
    FileSignature {
        name: "SQLite",
        magic: b"SQLite format 3",
        offset: 0,
    },
    FileSignature {
        name: "MP3/ID3",
        magic: b"ID3",
        offset: 0,
    },
    FileSignature {
        name: "RIFF/AVI/WAV",
        magic: b"RIFF",
        offset: 0,
    },
    FileSignature {
        name: "Mach-O 64",
        magic: b"\xCF\xFA\xED\xFE",
        offset: 0,
    },
    FileSignature {
        name: "Mach-O 32",
        magic: b"\xCE\xFA\xED\xFE",
        offset: 0,
    },
];

/// Scan a device for file signatures.
pub fn scan_signatures(
    device: &mut dyn RawDeviceIo,
    block_size: usize,
) -> Result<Vec<FileSignatureHit>> {
    let capacity = device.capacity();
    let mut buf = vec![0u8; block_size];
    let mut hits = Vec::new();

    const MAX_HITS: usize = 10_000;

    // Sample blocks across the device
    let total_blocks = capacity / block_size as u64;
    let step = if total_blocks > 4096 {
        total_blocks / 4096
    } else {
        1
    };

    let mut block_idx: u64 = 0;
    'outer: while block_idx < total_blocks {
        let offset = block_idx * block_size as u64;
        if offset >= capacity {
            break;
        }

        let read_len = ((capacity - offset) as usize).min(block_size);
        match device.read_at(offset, &mut buf[..read_len]) {
            Ok(n) if n > 0 => {
                // Check each signature against the start of this block
                for sig in SIGNATURES {
                    if sig.offset + sig.magic.len() <= n
                        && &buf[sig.offset..sig.offset + sig.magic.len()] == sig.magic
                    {
                        if hits.len() >= MAX_HITS {
                            log::warn!(
                                "Maximum signature hits ({}) reached, stopping scan",
                                MAX_HITS
                            );
                            break 'outer;
                        }
                        hits.push(FileSignatureHit {
                            offset,
                            file_type: sig.name.to_string(),
                            magic_hex: sig
                                .magic
                                .iter()
                                .map(|b| format!("{b:02X}"))
                                .collect::<Vec<_>>()
                                .join(" "),
                            confidence: "high".to_string(),
                        });
                    }
                }
            }
            _ => break,
        }

        block_idx += step;
    }

    Ok(hits)
}
